### PR TITLE
Initial rust implementation

### PR DESCRIPTION
**Pull request by JamesFysh**
working implementation (passes all validation, enc/dec, shorten/recover unit-tests); probably room for improvement wrt comments, layout.

### CODE REVIEW

File: rust/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "olc"
+description = "Library for translating between GPS coordinates (WGS84) and Open Location Code"
+version = "0.1.0"
+authors = ["James Fysh <james.fysh@gmail.com>"]
+license = "MIT/Apache-2.0"
```
Comment (drinckes) [inline rust/Cargo.toml]: Can you license this under just Apache-2.0 to match the rest of the project?

File: rust/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+repository = "https://github.com/JamesFysh/open-location-code"
```
Comment (drinckes) [inline rust/Cargo.toml]: Change to point to this repo

File: rust/src/interface.rs [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+/// 10 characters, returning a code of approximately 13.5x13.5 meters. Longer
+/// codes represent smaller areas, but lengths > 14 are sub-centimetre and so
+/// 11 or 12 are probably the limit of useful codes.
+pub fn encode(pt: Point<f64>, code_length: usize) -> String {
+    let mut lat = clip_latitude(pt.lat());
+    let mut lng = normalize_longitude(pt.lng());
+
+    // Latitude 90 needs to be adjusted to be just less, so the returned code
+    // can also be decoded.
+    if lat == LATITUDE_MAX {
```
Comment (drinckes) [inline rust/src/interface.rs]: Using == to compare floats can be optimistic. Maybe use a comparison that the difference between them is very, very small, or that lat is > LATITUDE_MAX? e.g.
`if lat > LATITUDE_MAX || (LATITUDE_MAX - lat) < 0.0000000001f64 {`

File: rust/tests/all_test.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        let lathi = cols[5].parse::<f64>().unwrap();
+        let lnghi = cols[6].parse::<f64>().unwrap();
+
+        let codearea = decode(code).unwrap();
+        assert_eq!(
+            encode(Point::new(lng, lat), codearea.code_length),
+            code,
+            "encoding lat={},lng={}", lat, lng
+        );
+        assert!((latlo - codearea.south).abs() < 0.001f64);
```
Comment (drinckes) [inline rust/tests/all_test.rs]: The difference should be < 1e-10.

File: rust/src/private.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        *lat *= GRID_ROWS;
+        *lng *= GRID_COLUMNS
+    }
+    if near(*lat) {
+        *lat = lat.round();
+    }
+    if near(*lng) {
+        *lng = lng.round();
+    }
+}
```
Comment (drinckes) [inline rust/src/private.rs]: Add a newline.

File: rust/src/private.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+pub fn near(value: f64) -> bool {
+    value.trunc() != (value + 0.0000000001f64).trunc()
+}
+
+pub fn narrow_region(digit: usize, lat: &mut f64, lng: &mut f64) {
+    if digit == 0 {
+        *lat /= ENCODING_BASE;
+        *lng /= ENCODING_BASE;
+    } else if digit < 10 {
```
Comment (drinckes) [inline rust/src/private.rs]: Replace 10 with PAIR_CODE_LENGTH?

File: rust/src/interface.rs [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        narrow_region(digit, &mut lat, &mut lng);
+
+        let lat_digit = lat as usize;
+        let lng_digit = lng as usize;
+        if digit < 10 {
+            code.push(CODE_ALPHABET[lat_digit]);
+            code.push(CODE_ALPHABET[lng_digit]);
+            digit += 2;
+        }
+        else {
```
Comment (drinckes) [inline rust/src/interface.rs]: Combine with previous line i.e.
} else {

File: rust/src/codearea.rs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    pub fn merge(self, other: CodeArea) -> CodeArea {
+        CodeArea::new(
+            self.south + other.south,
+            self.west + other.west,
+            self.north + other.north,
+            self.east + other.east,
+            self.code_length + other.code_length
+        )
+    }
+}
```
Comment (drinckes) [inline rust/src/codearea.rs]: Add newline.

File: rust/src/consts.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+];
+
+// Number of columns in the grid refinement method.
+pub const GRID_COLUMNS: f64 = 4f64;
+
+// Number of rows in the grid refinement method.
+pub const GRID_ROWS: f64 = 5f64;
+
+// Minimum length of a code that can be shortened.
+pub const MIN_TRIMMABLE_CODE_LEN: usize = 6;
```
Comment (drinckes) [inline rust/src/consts.rs]: Add newline.

File: rust/src/interface.rs [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+        latitude += area_range;
+    }
+    let longitude_diff = longitude - normalize_longitude(ref_pt.lng());
+    if longitude_diff > area_edge {
+        longitude -= area_range;
+    } else if longitude_diff < -area_edge {
+        longitude += area_range;
+    }
+    Ok(encode(Point::new(longitude, latitude), code_area.code_length))
+}
```
Comment (drinckes) [inline rust/src/interface.rs]: Add newline.

File: rust/tests/csv_reader.rs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        // Iterate lines in the CSV file, dropping empty & comment lines
+        while let Some(Ok(s)) = self.iter.next() {
+            if s.is_empty() || s.starts_with("#") {
+                continue;
+            }
+            return Some(s);
+        }
+        None
+    }
+}
```
Comment (drinckes) [inline rust/tests/csv_reader.rs]: Add newline.

File: rust/tests/all_test.rs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        let _short = cols[2] == "true";
+        let _full = cols[3] == "true";
+
+        assert_eq!(is_valid(code), _valid, "valid for code: {}", code);
+        assert_eq!(is_short(code), _short, "short for code: {}", code);
+        assert_eq!(is_full(code), _full, "full for code: {}", code);
+
+        tested += 1;
+    }
+    assert!(tested > 0);
```
Comment (drinckes) [inline rust/tests/all_test.rs]: Nice check ;-)

File: run_tests.sh [hunk 13]
```diff
@@ -32,4 +32,12 @@ if [ "$TEST_DIR" == "ruby" ]; then
   exit
 fi
 
+if [ "$TEST_DIR" == "rust" ]; then
+  curl https://sh.rustup.rs -sSf | sh -s -- -y
```
Comment (JamesFysh) [inline run_tests.sh]: Well, that was painful to implement!  I'll have to invest a bit of time in a local setup of TravisCI so I can verify it works before pushing changes, going forward!

Comment (drinckes) [inline run_tests.sh]: Yeah I know, Travis changes can be painful.

**Comment by JamesFysh** [comment]
I think I've covered off all the changes you asked for, and (finally) managed to get Travis CI working correctly.  Do I need to do anything to confirm that the requested changes have been made?

**Comment by drinckes** [comment]
Thanks @JamesFysh, you don't have to do anything, I don't have time today but I'll merge it early next week. Thanks!

**Pull request merged**